### ISSUE TITLE
Issue Using BatchNormalizationLayer 

### ISSUE THREAD

**Issue by nwindt** [ARM-software/ComputeLibrary#235]
I have been reimplementing layers of the Darknet framework using ACL. I have had trouble adding the BatchNormalizationLayer. It appears to me that BatchNormalizationLayer computes incorrect values for input tensors with a feature-map >= 256.

 Comparing my outputs of my ACL implemented BatchNormLayer to expected outputs, I can see my implementation worked with input tensors of the following size,  416X416X3, 208x208x16, 104X104X32, 52X52X64, 26x26x128. However got completely wrong outputs for input tensors with a large FM dimension: 13x13X256, 13x13x512, 13x13x1024. 

My implementation can be seen below

```
void CONV::BatchNormalization( float *output, int out_H, int out_W, int m, layer l)
{
	float  epsilon = .000001;
	int n = out_H * out_W;
	const TensorShape _filter_shape (m);
	const TensorShape _BatchnormShape (out_W, out_H, m);

	_BatchnormOut.allocator()->init(TensorInfo(_BatchnormShape, 1, DataType::F32));
	_BatchnormIn.allocator()->init(TensorInfo(_BatchnormShape, 1, DataType::F32));
	_mean.allocator()->init(TensorInfo(_filter_shape, 1, DataType::F32));
	_var.allocator()->init(TensorInfo(_filter_shape, 1, DataType::F32));
	_beta.allocator()->init(TensorInfo(_filter_shape, 1, DataType::F32));
	_gamma.allocator()->init(TensorInfo(_filter_shape, 1, DataType::F32));

	BatchNorm.configure(&_BatchnormIn, &_BatchnormOut, &_mean, &_var, &_beta, &_gamma, epsilon);
	_BatchnormIn.allocator()->allocate();
	_BatchnormOut.allocator()->allocate();
	_mean.allocator()->allocate();
	_var.allocator()->allocate();
	_beta.allocator()->allocate();
	_gamma.allocator()->allocate();

        memcpy( reinterpret_cast<float*>( _BatchnormIn.buffer() + _BatchnormIn.info()->offset_element_in_bytes(Coordinates(0,0,0))), output, n * m * sizeof( float ) );
	memcpy( reinterpret_cast<float*>( _mean.buffer() + _mean.info()->offset_element_in_bytes(Coordinates(0))), l.rolling_mean, m * sizeof( float ) );
	memcpy( reinterpret_cast<float*>( _var.buffer() + _var.info()->offset_element_in_bytes(Coordinates(0))), l.rolling_variance, m * sizeof( float ) );
	memcpy( reinterpret_cast<float*>( _beta.buffer() + _beta.info()->offset_element_in_bytes(Coordinates(0))), l.biases, m * sizeof( float ) );
	memcpy( reinterpret_cast<float*>( _gamma.buffer() + _gamma.info()->offset_element_in_bytes(Coordinates(0))), l.scales, m * sizeof( float ) );


    arm_compute::Scheduler::get().set_num_threads(8);

    start_time = boost::posix_time::microsec_clock::local_time();
	BatchNorm.run();
	end_time = boost::posix_time::microsec_clock::local_time();

	dur = end_time - start_time;
	std::cout << "BatchNorm took: " << dur.total_milliseconds() << " ms" << std::endl;

	memcpy( output, reinterpret_cast<float *>( _BatchnormOut.buffer() + _BatchnormOut.info()->offset_element_in_bytes(Coordinates(0,0,0))), m * n * sizeof( float ) ); 
}
```

**Comment by mpekatsoula** [comment]
It looks like when you copy the data back you don't take the padding into account. [Here](https://arm-software.github.io/ComputeLibrary/v17.09/architecture.xhtml#S4_6_4_working_with_objects) you can find examples of how to work with tensors.

**Comment by nwindt** [comment]
@mpekatsoula, Copying the tensors a row at a time fixed my issue. In this case, why can I not do a direct memcpy? how does this relate to padding?

I am now trying to pass the output tensor of a ConvolutionLayer function directly to a BatchNormLayer function. In some cases the last two columns of each row are being computed to 0 by the BatchNorm function.  Any idea what would cause this issue?  I am able to work around this by copying out the values of the Convolution output tensor and recopying them to a new tensor which I pass to the BatchNormLayer.

**Comment by GeorgeARM** [comment]
Hello @nwindt,

We use vector operations to optimize our algorithms. Thus, when the width of a tensor is not a multiple of the vector size that we use then we usually pad the tensor to avoid having special kernels or branching for handling the leftover pixels. This paddidn is added during the configuration of the function, that's why we usually suggest to do the allocation of the tensors after the configuration of the functions.

So the memcpy worked for some of your tensor inputs because their width was possibly a multiple of the vector size that is used by the batch normalization function (I think that it is 4 for float input) so no padding was added. On the other hand, if padding is added you cannot use a memcpy directly, as there is padding at the end of each row, and you will copy values in the padding pixels.

**Comment by nwindt** [comment]
Thank you @GeorgeARM.
